File: parse/cat.py
from .parse import Parse, Node

# cats multiple parses together
def cat(*parsers):
	return ParseCat(parsers)

# parses only if all parsers when catted together parse
class ParseCat(Parse):
	def __init__(self, parsers):
		self.parsers = parsers

	def parse(self, file, i):
		start = i
		results = []

		for parser in self.parsers:
			result = parser.parse(file, i)
			if result == None:
				return None

			results.append(result)
			i = result.end
		return Cat(result, start, i, file)

class Cat(Node):
	def __init__(self, nodes, start, end, file):
		Node.__init__(self, start, end, file)
		self.nodes = nodes

File: parse/cat.lua
-- (...) are parsers, ie magic fns
-- that expect (file,i)
-- parses only if they parse in sequence
function cat(file, i, ...)
	local parsers = table.pack(...)
	local j = i
	elems = {}
	for _,parser in ipairs(parsers) do
		node = parser(file, j)()
		if node == nil then
			return nil
		end

		table.insert(elems, node)
		j = node.j
	end
	return table.unpack(elems)
end